src/library/sidebarmodel.cpp

```diff
@@ -608,3 +599,303 @@ void SidebarModel::slotFeatureSelect(LibraryFeature* pFeature, const QModelIndex
+        featureRow = m_sFeatures.indexOf(pFeature);
+    }
+    TreeItemModel* pChildModel = pFeature->sidebarModel();
+    DEBUG_ASSERT(pChildModel);
+    TreeItem* pBaseItem = pChildModel->getRootItem();
+
+    qWarning() << " >> check bookmark children of" << pFeature->title().toString();
+
+    bool updated = false;
+    // QList<SidebarBookmark> missingBookmarks;
+    for (auto& bm : m_bookmarks) {
+        if (bm.featureRow != featureRow) {
+            continue;
+        }
+        // Lookup bookmark.
+        // If found, store index. Update bookmark childlevel and parentRow?
+        // Else remove bookmark and index from list.
+        const auto bmIndex = findBookmarkIndex(bm, pBaseItem);
+        if (bmIndex.isValid()) {
+            qWarning() << "   >> update" << bm.data << bm.label;
+            // TODO Use createBookmarkFromIndex() and swap with bm in order to
+            // update bm.childLevel and bm.parentRow, too?
+            bm.index = bmIndex;
+            if (!updated) {
+                updated = true;
+            }
+        } else {
+            qWarning() << "   >> no match for" << bm.data << bm.label;
+            // missingBookmarks.append(bm);
+        }
+    }
+
+    if (!updated) {
+        // no hit for affected feature, nothing to do
+        return;
+    }
+
+    // For now don't remove missing bookmarks.
+    // BrowseFeature bookmarks may be 'missing' after collapsing and
+    // re-expanding a directory tree one or more levels above a bookmark.
+    // Reason is that on expand only the next sublevel is rebuilt, so
+    // and bookmarked items on lower levels are simply not there, yet.
+    // for (auto& bm : missingBookmarks) {
+    //        qWarning() << "   >> remove" << bm.data << bm.label;
+    //    m_bookmarks.removeOne(bm);
+    // }
+
+    sortBookmarksUpdateIndices();
+}
+
+/// Try to find the TreeItem in a feature's childmodel that matches the bookmark.
+/// Return its index when found, else return invalid QModelIndex().
+/// Scans the entire tree recursively, either by item data or label.
+// TODO Do we need to store an index list at all when using match()?
+// Check performance!
+QModelIndex SidebarModel::findBookmarkIndex(
+        const SidebarBookmark& bookmark,
+        const TreeItem* pItem) {
+    // qWarning() << " findBookmarkIndex"
+    // qWarning() << "   data: " << pItem->getData();
+    // qWarning() << "   curr ch level: " << pItem->childLevel();
+    // qWarning() << "   bookm ch level:" << bookmark.childLevel;
+
+    auto pFeature = pItem->feature();
+    TreeItemModel* pChildModel = pFeature->sidebarModel();
+    DEBUG_ASSERT(pChildModel);
+    QModelIndexList results;
+    if (pFeature->isItemDataUnique(bookmark.data)) {
+        results = pChildModel->match(
+                pChildModel->getRootIndex(),
+                TreeItemModel::kDataRole,
+                bookmark.data,
+                1,
+                Qt::MatchWrap | Qt::MatchExactly | Qt::MatchRecursive);
```

Comment:
I discovered by chance `QAbstractItemModel::match()` **after** I wrote my own recursive match() function that checked label and/or data :grimacing: :man_facepalming:  :laughing: